pkg/dns/aws/dns.go

```diff
@@ -167,6 +181,9 @@ func NewProvider(config Config, operatorReleaseVersion string) (*Provider, error
 				r53Config = r53Config.WithEndpoint(ep.URL)
 				log.Info("using route53 custom endpoint", "url", ep.URL)
 			case ep.Name == TaggingService:
+				if tagConfig == nil {
+					continue
+				}
```

Comment:
I am fine with logging a warning if you like. The user should never specify a custom service endpoint for the resourcegrouptagging API when running in a C2S region, however. I contend that warning about the service endpoint is not terribly helpful since the ingress operator will not be accessing that endpoint regardless.

As an aside, the logic here for the custom endpoints should be taking into consideration the region specified for the custom endpoints. For example, the custom endpoints could contain a specification for the endpoint to use when accessing us-west-1, which should not be considered if the region installed to is actually us-east-1. Before using a custom endpoint, there should be a check that the region of the service endpoint matches the region of the installation.

Reply:
> The user should never specify a custom service endpoint for the resourcegrouptagging API when running in a C2S region, however. I contend that warning about the service endpoint is not terribly helpful since the ingress operator will not be accessing that endpoint regardless.

Maybe the user should never, but it looks like the user can.  The relevant API godoc just says, "serviceEndpoints list contains custom endpoints which will override default service endpoint of AWS Services."  If the user doesn't specify custom endpoints, the operator figures out which endpoints to use based on the region, but if the user _does_ specify custom endpoints, it seems like operators must use those custom endpoints to comply with the API.  That said, I did not write the API, I am only trying to interpret it, and I'll defer your judgment here.  

> As an aside, the logic here for the custom endpoints should be taking into consideration the region specified for the custom endpoints. For example, the custom endpoints could contain a specification for the endpoint to use when accessing us-west-1, which should not be considered if the region installed to is actually us-east-1. Before using a custom endpoint, there should be a check that the region of the service endpoint matches the region of the installation.

How can the user specify a region for a custom endpoint?  The API has no field to specify an endpoint's region: 
```go
// AWSServiceEndpoint store the configuration of a custom url to
// override existing defaults of AWS Services.
type AWSServiceEndpoint struct {
	// name is the name of the AWS service.
	// The list of all the service names can be found at https://docs.aws.amazon.com/general/latest/gr/aws-service-information.html
	// This must be provided and cannot be empty.
	//
	// +kubebuilder:validation:Pattern=`^[a-z0-9-]+$`
	Name string `json:"name"`

	// url is fully qualified URI with scheme https, that overrides the default generated
	// endpoint for a client.
	// This must be provided and cannot be empty.
	//
	// +kubebuilder:validation:Pattern=`^https://`
	URL string `json:"url"`
}
```
https://github.com/openshift/api/blob/bb6857301efec6d005852debef3030781c26dd78/config/v1/types_infrastructure.go#L239-L255

Reply:
Regarding the region, I don't know the history here, but I would have expected the operator to read the service endpoints from the cloud config rather than from the Infrastructure resource. The service endpoints in the Infrastructure resource are inputs that the cluster-config-opreator uses to generate the cloud config. In the cloud config, each service endpoint is specific to a region.

https://github.com/kubernetes/kubernetes/blob/35bf00acb1f5f2be22cdcb180c5e1227a72934fc/staging/src/k8s.io/legacy-cloud-providers/aws/aws.go#L626-L646

Reply:
> if the user does specify custom endpoints, it seems like operators must use those custom endpoints to comply with the API

I think that we are compliant with that. We use the custom endpoint specified for every call to the resourcegroupstagging API. In the case of C2S, we do not make any calls to the resourcegroupstagging API.